File: haohaninfo/回測範例/46.py
# -*- coding: UTF-8 -*-
#取I020，依照逗點分隔，並將分隔符號去除
I020 = [ line.strip('\n').split(",") for line in open('Futures_20170815_I020.csv')][1:]

#起始時間至結束時間
I020a= [ int(line[4]) for line in I020 if int(line[0])<=9000000 ]
I020b= [ line for line in I020 if int(line[0])>9000000 and int(line[0])<11000000 ]

#定義上下界
ceil=max(I020a)
floor=min(I020a)
#倉位為0
index=0

for i in range(len(I020b)):
 price=int(I020b[i][4])
 #進場判斷
 if index==0:
  if price>ceil:
   OrderTime=I020b[i][0]			#新倉時間紀錄
   OrderPrice=price 				#新倉價格紀錄
   index=1
   print "Buy OrderTime:",OrderTime," OrderPrice:",OrderPrice,
  elif price<floor:
   OrderTime=I020b[i][0]			#新倉時間紀錄
   OrderPrice=price 				#新倉價格紀錄
   index=-1
   print "Sell OrderTime:",OrderTime," OrderPrice:",OrderPrice,
  elif i == len(I020b)-1:
   print "No Trade"	
   break
 #出場判斷
 elif index!=0:
  if index==1:
   if OrderPrice+20<=price or OrderPrice-10>=price:
    CoverTime=I020b[i][0]  			#平倉時間紀錄
    CoverPrice=int(I020b[i][4]) 	#平倉時間紀錄
    print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",CoverPrice-OrderPrice
    break
   elif i == len(I020b)-1:
    CoverTime=I020b[i][0]  			#平倉時間紀錄
    CoverPrice=int(I020b[i][4])  	#平倉時間紀錄
    print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",CoverPrice-OrderPrice
  elif index==-1:
   if price<=OrderPrice-20 or price>=OrderPrice+10:
    CoverTime=I020b[i][0]  			#平倉時間紀錄
    CoverPrice=int(I020b[i][4])  	#平倉時間紀錄
    print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",OrderPrice-CoverPrice
    break
   elif i == len(I020b)-1:
    CoverTime=I020b[i][0]  			#平倉時間紀錄
    CoverPrice=int(I020b[i][4])  	#平倉時間紀錄
    print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",OrderPrice-CoverPrice

File: haohaninfo/交易回測案例/46.R
I020 <- read.csv('Futures_20170815_I020.csv')

# 選擇必要欄位
A01 <- I020[c(1,5,6)]
names(A01) <- c("INFO_TIME","PRICE","QTY") 

# 選擇趨勢時段
A02 <- subset(A01, INFO_TIME>=as.numeric(08450000) & INFO_TIME<=as.numeric(09000000)) 
MaxPrice <- max(A02$PRICE)
MinPrice <- min(A02$PRICE)
Spread <- MaxPrice - MinPrice 

# 起始及出場時段
A03 <- subset(A01, INFO_TIME>as.numeric(09000000) & INFO_TIME<=as.numeric(10000000)) 

Index<-0

# 開始進行回測
for(i in 1:nrow(A03)){
 #判斷進倉
 if(Index==0 & A03[i,]$PRICE>MaxPrice+Spread*as.numeric(0.2)){
  Index=1 
  OrderTime <- A03[i,]$INFO_TIME
  OrderPrice <- A03[i,]$PRICE   
  #cat(Time0,"BUY",OrderPrice,"StopPoint",StopPoint,"\n")
 }else if (Index==0 & A03[i,]$PRICE<MinPrice-Spread*as.numeric(0.2)){
  Index=-1
  OrderTime <- A03[i,]$INFO_TIME
  OrderPrice <- A03[i,]$PRICE   
  #cat(Time0,"SELL",OrderPrice,"StopPoint",StopPoint,"\n")
 }
 # 判斷平倉
 if(Index==1){
  if((A03[i,]$PRICE > OrderPrice+20 || A03[i,]$PRICE < OrderPrice-10)){
   CoveryTime <- A03[i,]$INFO_TIME
   CoveryPrice <- A03[i,]$PRICE
   cat("BUY TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",CoveryPrice-OrderPrice,'\n') 
   break
  }else if(i==nrow(A03)){
   CoveryTime <- A03[i,]$INFO_TIME
   CoveryPrice <- A03[i,]$PRICE
   cat("BUY TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",CoveryPrice-OrderPrice,'\n')  
  }
 }else if(Index==-1){
  if((A03[i,]$PRICE < OrderPrice-20 || A03[i,]$PRICE > OrderPrice+10)){
   CoveryTime <- A03[i,]$INFO_TIME
   CoveryPrice <- A03[i,]$PRICE
   cat("SELL TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",OrderPrice-CoveryPrice,'\n')
   break
  }else if(i==nrow(A03)){
   CoveryTime <- A03[i,]$INFO_TIME
   CoveryPrice <- A03[i,]$PRICE
   cat("SELL TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",OrderPrice-CoveryPrice,'\n')
  }
 }  
}

